include/drivers/counter.h

```diff
@@ -460,6 +501,72 @@ static inline u32_t z_impl_counter_get_max_relative_alarm(struct device *dev)
 	return api->get_max_relative_alarm(dev);
 }
 
+/**
+ * @brief Set guard period in counter ticks.
+ *
+ * Setting non-zero guard period enables detection of setting absolute alarm
+ * too late. It limits how far in the future absolute alarm can be set.
+ *
+ * Detection of too late setting is vital since if it is not detected alarm
+ * is delayed by full period of the counter (up to 32 bits). Because of the
+ * wrapping, it is impossible to distinguish alarm which is short in the past
+ * from alarm which is targeted to expire after full counter period. In order to
+ * detect too late setting, longest possible alarm is limited. Absolute value
+ * cannot exceed: (now + top_value - guard_period) % top_value.
+ *
+ * Guard period depends on application and counter frequency. If it is expected
+ * that absolute alarms setting might be delayed then guard period should
+ * exceed maximal potential delay. If use case allows, guard period can be set
+ * very high (e.g. half of the counter top value).
+ *
+ * After initialization guard period is set to 0 and late detection is disabled.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param ticks		Guard period in counter ticks.
+ * @param flags		See @ref COUNTER_GUARD_PERIOD_FLAGS.
+ *
+ * @retval 0 if successful.
+ * @retval -ENOTSUP if function or flags are not supported.
+ * @retval -EINVAL if ticks value is invalid.
+ */
+__syscall int counter_set_guard_period(struct device *dev, u32_t ticks,
+					u32_t flags);
+
+static inline int z_impl_counter_set_guard_period(struct device *dev,
+						   u32_t ticks, u32_t flags)
+{
+	const struct counter_driver_api *api =
+				(struct counter_driver_api *)dev->driver_api;
+
+	if (!api->set_guard_period) {
```

Comment:
This isn't typical of Zephyr driver subsystems, normally if an API isn't implemented we don't do a NULL check. We might want to consider a consistent system-wide strategy on this.

Reply:
this approach allows extending api without breaking existing drivers. It's not the first place it is used though i'm not sure how consistent zephyr is in that matter.

Reply:
> i'm not sure how consistent zephyr is in that matter.

It's not, AFAICT.
Fine with this as-is, can address kernel-wide later which direction to go (and codify somewhere)